torch/utils/cpp_extension.py

```diff
@@ -538,29 +538,38 @@ def build_extensions(self) -> None:
         compiler_name, compiler_version = self._check_abi()
 
         cuda_ext = False
+        sycl_ext = False
         extension_iter = iter(self.extensions)
         extension = next(extension_iter, None)
-        while not cuda_ext and extension:
+        while (not cuda_ext and not sycl_ext) and extension:
             for source in extension.sources:
                 _, ext = os.path.splitext(source)
                 if ext == '.cu':
                     cuda_ext = True
+                elif ext == '.sycl':
+                    sycl_ext = True
+
+                if cuda_ext and sycl_ext:
                     break
+
             extension = next(extension_iter, None)
 
+        if sycl_ext:
+            assert self.use_ninja, "ninja is required to build sycl extensions."
+
         if cuda_ext and not IS_HIP_EXTENSION:
             _check_cuda_version(compiler_name, compiler_version)
 
         for extension in self.extensions:
-            # Ensure at least an empty list of flags for 'cxx' and 'nvcc' when
+            # Ensure at least an empty list of flags for 'cxx', 'nvcc' and 'sycl' when
```

Comment:
The list is a compiler list. In terms of SYCL compiler, Intel's brand is DPC++. @gujinghui , do you think we should use `dpc++` here rather than `sycl`?

Reply:
Note that here `sycl` is a key of `extra_compile_args` dictionary which sets a mapping between languages and additional compiler flags to pass to compiler. Note comment to this function:

https://github.com/pytorch/pytorch/blob/729b7c0a84f8819e0bb31ce8f1e048d7bde64b45/torch/utils/cpp_extension.py#L500-L503

I have introduced a new key called `sycl` for the reason that I am actually trying to abstract away from Intel branded DPC++ compiler. Note that in the future. hopefully, SYCL support will be available in clang. And in general there might be other SYCL compilers besides DPC++. Thus, question here is: should they key be named per language (in this case it should be `sycl`) or per-compiler (in this case I should rename to `dpc++`)?

Reply:
In this context, `cxx` and `nvcc` are specifying the compilers, instead of languages.
If not introduce `dpc++`, we should use `sycl compiler` at least, instead of `sycl`?

Reply:
> If not introduce `dpc++`, we should use `sycl compiler` at least, instead of sycl?

`sycl compiler` does not make a sense to me as a dictionary key. Review https://github.com/pytorch/pytorch/blob/main/test/cpp_extensions/setup.py or here in this PR above. `nvcc`, `cxx` and (this PR introduces) `sycl` are dictionary keys:

```
    extension = CUDAExtension(
        "torch_test_cpp_extension.torch_library",
        ["torch_library.cu"],
        extra_compile_args={"cxx": CXX_FLAGS, "nvcc": ["-O2"]},
    )
...
    extension = SyclExtension(
        "torch_test_cpp_extension.sycl",
        ["xpu_extension.sycl"],
        extra_compile_args={"cxx": CXX_FLAGS, "sycl": ["-O2"]},
    )
```